pkg/kafka/partitionring/consumer/balancer.go

```diff
@@ -0,0 +1,159 @@
+package consumer
+
+import (
+	"sort"
+
+	"github.com/grafana/dskit/ring"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/kmsg"
+)
+
+type cooperativeActiveStickyBalancer struct {
+	kgo.GroupBalancer
+	partitionRing ring.PartitionRingReader
+}
+
+// NewCooperativeActiveStickyBalancer creates a balancer that combines Kafka's cooperative sticky balancing
+// with partition ring awareness. It works by:
+//
+// 1. Using the partition ring to determine which partitions are "active" (i.e. should be processed)
+// 2. Filtering out inactive partitions from member assignments during rebalancing, but still assigning them
+// 3. Applying cooperative sticky balancing only to the active partitions
+//
+// This ensures that:
+// - Active partitions are balanced evenly across consumers using sticky assignment for optimal processing
+// - Inactive partitions are still assigned and consumed in a round-robin fashion, but without sticky assignment
+// - All partitions are monitored even if inactive, allowing quick activation when needed
+// - Partition handoff happens cooperatively to avoid stop-the-world rebalances
+//
+// This balancer should be used with [NewGroupClient] which monitors the partition ring and triggers
+// rebalancing when the set of active partitions changes. This ensures optimal partition distribution
+// as the active partition set evolves.
+func NewCooperativeActiveStickyBalancer(partitionRing ring.PartitionRingReader) kgo.GroupBalancer {
+	return &cooperativeActiveStickyBalancer{
+		GroupBalancer: kgo.CooperativeStickyBalancer(),
+		partitionRing: partitionRing,
+	}
+}
+
+func (*cooperativeActiveStickyBalancer) ProtocolName() string {
+	return "cooperative-active-sticky"
+}
+
+func (b *cooperativeActiveStickyBalancer) MemberBalancer(members []kmsg.JoinGroupResponseMember) (kgo.GroupMemberBalancer, map[string]struct{}, error) {
+	// Get active partitions from ring
+	activePartitions := make(map[int32]struct{})
+	for _, id := range b.partitionRing.PartitionRing().PartitionIDs() {
```

Comment:
Should this be ActivePartitionIDs?
I'm unsure how we want to assign Inactive partitions (i.e. the ones where we've stopped writing data but some old data is still available)

Reply:
I decide to only move partition when no data is required so I also consider inactive as active.